src/index.ml

```diff
@@ -708,8 +708,10 @@ struct
         Thread.return ()
     | Some witness -> merge ?hook ~witness t
 
-  let replace t key value =
+  let replace_with_timer ?with_timer t key value =
```

Comment:
I feel that `replace_with_timer` should probably be implemented in terms of `replace` and not vice-versa. That way, this timing logic isn't touched in production and `?with_timer` doesn't need to be optional. I'm also suspicious of starting the timer after some of `replace` has already happened, in this case the `check_open`, and with the other implementation this can never occur.

Reply:
Also, is it necessary to have the sampling logic inside Index at all? AFAICS, `replace` is only ever called externally. Perhaps it would be better to have `replace_timed : t -> k -> v -> <time interval>` and keep sampling considerations in the benchmarks.

Reply:
> I feel that replace_with_timer should probably be implemented in terms of replace and not vice-versa. 

Yes, good point, thank you. 

> Also, is it necessary to have the sampling logic inside Index at all? AFAICS, replace is only ever called externally.

I'm not sure about this. It can be called from irmin, and the [initial motivation](https://ocamllabs.slack.com/archives/GNZ0111EH/p1585229720001200) for this was, I think to have these stats from tezos. I would like to have the sampling logic somewhere, to be used outside of our benchmarks.

Reply:
> I'm not sure about this. It can be called from irmin, and the initial motivation for this was, I think to have these stats from tezos. I would like to have the sampling logic somewhere, to be used outside of our benchmarks.

That's fair. In that case, I'd like some other name than `with_timer` (`sampling_interval`, perhaps?).